src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -612,23 +615,44 @@ public override bool Equals(object comparand)
 
         public override int GetHashCode()
         {
-            // For IPv6 addresses, we cannot simply return the integer
-            // representation as the hashcode. Instead, we calculate
-            // the hashcode from the string representation of the address.
+            if (_hashCode != 0)
+            {
+                return _hashCode;
+            }
+
+            // For IPv6 addresses, we calculate the hashcode by using Marvin
+            // on a stack-allocated array containing the Address bytes and ScopeId.
+            int hashCode;
             if (IsIPv6)
             {
-                if (_hashCode == 0)
+                const int addressAndScopeIdLength = IPAddressParserStatics.IPv6AddressBytes + sizeof(uint);
+                Span<byte> addressAndScopeIdSpan;
+                unsafe
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    byte* addressAndScopeId = stackalloc byte[addressAndScopeIdLength];
+                    addressAndScopeIdSpan = new Span<byte>(addressAndScopeId, addressAndScopeIdLength);
```

Comment:
corefx is using a preview 7.2 compiler that has additional span-related features.  This can be changed to just:
```C#
Span<byte> addressAndScopeIdSpan = stackalloc byte[addressAndScopeIdLength];
```